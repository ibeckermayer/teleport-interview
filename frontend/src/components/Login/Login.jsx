import React from 'react';
import { useHistory } from 'react-router-dom';

const Login = () => {
  const history = useHistory();

  const tryLogin = async e => {
    // Form validation is handled by html5
    e.preventDefault();
    const res = await fetch('api/login', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        email: e.target.email.value,
        password: e.target.password.value,
      }),
    });
    if (res.ok) {
      history.push('/dashboard');
    }
    // TODO: check for Unauthorized and alert user that username/pwd is incorrect
  };

  return (
    <form className="login-form" onSubmit={tryLogin}>
      <h1>Sign Into Your Account</h1>

      <div>
        <label htmlFor="email">Email Address</label>
        <input
          type="email"
          id="email"
          className="field"
          autoComplete="username"
          name="email"
          required
        />
      </div>

      <div>
        <label htmlFor="password">Password</label>
        <input
          type="password"
          id="password"
          className="field"
          autoComplete="current-password"
          name="password"
          required
        />
      </div>

      <input
        type="submit"
        value="Login to my Dashboard"
        className="button block"
      />
    </form>
  );
};

export default Login;
