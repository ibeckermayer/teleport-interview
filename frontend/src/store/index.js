import { StoreContext, AppContext } from './store';

export { StoreContext, AppContext };
