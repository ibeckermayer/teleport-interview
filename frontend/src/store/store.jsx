import React, { createContext, useReducer, useEffect } from 'react';
import { getLocalStore, setLocalStore } from '../localStorage';

// Default state for the store
const initialStore = {
  sessionID: '', // Empty string represents user not authenticated, otherwise they are authenticated
  plan: '', // "FREE" or "ENTERPRISE"
};

// On app load, try to fetch a saved store from localStorage
const localStore = getLocalStore();

// Updates the store with newStoreValues
// Pass `newStoreValues = null` to reset the store
const storeReducer = (store, newStoreValues) => {
  if (newStoreValues === null) {
    localStorage.removeItem('store');
    return initialStore;
  }
  return { ...store, ...newStoreValues };
};

// Global React.Context value for components to pass into useContext in order to access the store
const StoreContext = createContext();

// createStore should wrap the entire application, so that any component in the application
// can access the global store's state as a StoreContext.Consumer
const createStore = WrappedComponent => props => {
  // Initialize with the store saved in localStorage if it exists, else initialize with initialStore
  const [store, dispatch] = useReducer(
    storeReducer,
    localStore || initialStore
  );

  // Save the store to localStorage every time it's changed
  useEffect(() => {
    setLocalStore(store);
  }, [store]);

  return (
    <StoreContext.Provider value={[store, dispatch]}>
      <WrappedComponent {...props} />
    </StoreContext.Provider>
  );
};

export { StoreContext, createStore };
