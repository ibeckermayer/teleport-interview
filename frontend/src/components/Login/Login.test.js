/**
 * @jest-environment jsdom
 */
import React from 'react';
import { mount } from 'enzyme';
import Login from './Login';
import { StoreContext } from '../../store';
import toJson from 'enzyme-to-json';

let store = {};
let setStore = newStoreValues => {
  const newStore =
    newStoreValues === null ? null : { ...store, ...newStoreValues };
  store = newStore;
};

test('it renders consistently', () => {
  const wrapper = mount(
    <StoreContext.Provider value={{ store, setStore }}>
      <Login />
    </StoreContext.Provider>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
