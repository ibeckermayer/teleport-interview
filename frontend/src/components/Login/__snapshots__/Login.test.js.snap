// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders consistently 1`] = `
<Login>
  <form
    className="login-form"
    onSubmit={[Function]}
  >
    <h1>
      Sign Into Your Account
    </h1>
    <div>
      <label
        htmlFor="email"
      >
        Email Address
      </label>
      <input
        autoComplete="username"
        className="field"
        id="email"
        name="email"
        required={true}
        type="email"
      />
    </div>
    <div>
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        autoComplete="current-password"
        className="field"
        id="password"
        name="password"
        required={true}
        type="password"
      />
    </div>
    <input
      className="button block"
      type="submit"
      value="Login to my Dashboard"
    />
  </form>
</Login>
`;
